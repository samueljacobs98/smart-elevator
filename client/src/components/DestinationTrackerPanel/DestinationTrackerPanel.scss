@use "../../assets/scss/variables" as *;

.destination-tracker-panel {
  background-color: $colour-grey-dark;
  padding: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  flex: 1;
  overflow-y: auto;
}
