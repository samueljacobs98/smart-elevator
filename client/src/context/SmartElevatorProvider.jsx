import { createContext, useState, useCallback } from "react";
import { getLiftStatus } from "../utils/api";
import Config from "../config";

export const SmartElevatorContext = createContext();

export const SmartElevatorProvider = ({ children }) => {
  const [showModal, setShowModal] = useState(false);
  const [liftStatus, setLiftStatus] = useState(null);
  const [floors, setFloors] = useState([]);
  const [queueData, setQueueData] = useState([]);
  const [arrived, setArrived] = useState(new Set());
  const [modalData, setModalData] = useState({
    lift: null,
    toFloor: null,
  });

  const getLiftStatusCallback = useCallback(async () => {
    const status = await getLiftStatus(Config.userFloor);
    if (status) setLiftStatus(status);
  }, [setLiftStatus]);

  return (
    <SmartElevatorContext.Provider
      value={{
        showModal,
        setShowModal,
        liftStatus,
        setLiftStatus,
        floors,
        setFloors,
        queueData,
        setQueueData,
        arrived,
        setArrived,
        modalData,
        setModalData,
        getLiftStatusCallback,
      }}
    >
      {children}
    </SmartElevatorContext.Provider>
  );
};
