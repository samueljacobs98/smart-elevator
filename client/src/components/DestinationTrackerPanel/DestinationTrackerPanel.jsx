import "./DestinationTrackerPanel.scss";
import { useContext } from "react";
import DestinationTrackerItem from "../DestinationTrackerItem/DestinationTrackerItem";
import { SmartElevatorContext } from "../../context/SmartElevatorProvider";
import Config from "../../config";

const DestinationTrackerPanel = () => {
  const { queueData, arrived } = useContext(SmartElevatorContext);

  const userFloor = Config.userFloor;

  return (
    <div className="destination-tracker-panel">
      {Object.entries(queueData).map(([floor, lifts]) => {
        if (parseInt(floor) === userFloor) return null;
        return (
          <DestinationTrackerItem
            key={floor}
            floor={floor}
            lifts={lifts}
            arrived={arrived}
          />
        );
      })}
    </div>
  );
};

export default DestinationTrackerPanel;
