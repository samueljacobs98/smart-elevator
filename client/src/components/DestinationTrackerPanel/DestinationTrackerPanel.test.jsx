import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import DestinationTrackerPanel from "./DestinationTrackerPanel";
import { SmartElevatorContext } from "../../context/SmartElevatorProvider";
import Config from "../../config";

jest.mock("../../config", () => ({
  userFloor: 2,
}));

jest.mock("../DestinationTrackerItem/DestinationTrackerItem", () => (props) => (
  <div data-testid="destination-tracker-item" {...props}>
    {props.floor}
  </div>
));

const mockQueueData = {
  1: ["lift 1"],
  2: ["lift 2"],
  3: ["lift 3"],
};
const mockArrived = new Set(["lift 1"]);

const renderComponent = () =>
  render(
    <SmartElevatorContext.Provider
      value={{ queueData: mockQueueData, arrived: mockArrived }}
    >
      <DestinationTrackerPanel />
    </SmartElevatorContext.Provider>
  );

describe("DestinationTrackerPanel", () => {
  it("renders without crashing", () => {
    renderComponent();

    expect(screen).toBeDefined();
  });

  it("renders a DestinationTrackerItem for each floor apart from the userFloor", () => {
    renderComponent();

    const trackerItems = screen.getAllByTestId("destination-tracker-item");

    expect(trackerItems.length).toBe(2);

    trackerItems.forEach((item) => {
      expect(item).not.toHaveTextContent(Config.userFloor);
    });
  });

  it("should match snapshot", () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });
});
