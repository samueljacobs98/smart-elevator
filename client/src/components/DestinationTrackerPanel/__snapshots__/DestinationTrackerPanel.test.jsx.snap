// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationTrackerPanel should match snapshot 1`] = `
<div>
  <div
    class="destination-tracker-panel"
  >
    <div
      arrived="[object Set]"
      data-testid="destination-tracker-item"
      floor="1"
      lifts="lift 1"
    >
      1
    </div>
    <div
      arrived="[object Set]"
      data-testid="destination-tracker-item"
      floor="3"
      lifts="lift 3"
    >
      3
    </div>
  </div>
</div>
`;
